### PR TITLE
[Merged by Bors] - light client optimistic update reprocessing

### PR DESCRIPTION
**Pull request by GeemoCandama**
## Issue Addressed
Currently there is a race between receiving blocks and receiving light client optimistic updates (in unstable), which results in processing errors. This is a continuation of PR #3693 and seeks to progress on issue #3651

## Proposed Changes

Add the parent_root to ReprocessQueueMessage::BlockImported so we can remove blocks from queue when a block arrives that has the same parent root. We use the parent root as opposed to the block_root because the LightClientOptimisticUpdate does not contain the block_root.

If light_client_optimistic_update.attested_header.canonical_root() != head_block.message().parent_root() then we queue the update. Otherwise we process immediately.
## Additional Info
michaelsproul came up with this idea.
The code was heavily based off of the attestation reprocessing.
I have not properly tested this to see if it works as intended.

### CODE REVIEW

**Comment by GeemoCandama** [comment]
It seems to be working properly to me. I added a couple of metrics to test similar to the attestation metrics. Sent Optimistic Update := moved to reprocessing queue during processing because it failed the check. Matched Optimistic Update := a gossiped block with a parent root matching at least one of the updates in the queue was imported (BlockImported). Expired Optimistic Update := an update with a matching parent root has not arrived within the QUEUED_LC_UPDATE_DELAY (12s) so we cleared that update from the queue. 

Is 12s a reasonable duration for QUEUED_LC_UPDATE_DELAY? I just copied the attestation delay.

I set MAXIMUM_QUEUED_LIGHT_CLIENT_UPDATES = 128 . What do you think of this value. It seems like it could be much lower if that is more desirable.

Screenshot from 2022-12-19 05-03-51

All errors except UnknownParentRoot are combined into the Error graph. OptimisticUpdateAlreadySeen is the only one I've gotten so far.

EDIT: I also cloned the optimistic update during processing because I wasn't sure how to satisfy the type system when having a LightClientOptimisticUpdate as a field of Error::UnknownParentRoot.

**Comment by GeemoCandama** [comment]
This is ready for review from my perspective. The issue in the EDIT portion of my previous comment still stands, and I'd appreciate your thoughts on this.

**Comment by michaelsproul** [comment]
Awesome thanks @GeemoCandama!

Cloning the optimistic update is probably fine. It _might_ be possible to work around it, but it's probably not a big deal if we can't.

I'm on leave until the 9th of Jan and I'll dig in for a proper review then

File: beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs [hunk 1]
```diff
@@ -107,6 +111,16 @@ impl<T: BeaconChainTypes> VerifiedLightClientOptimisticUpdate<T> {
             None => return Err(Error::SigSlotStartIsNone),
         }
 
+        // check if we can process the optimistic update immediately
+        // otherwise queue
+        let canonical_root = light_client_optimistic_update
+            .attested_header
+            .canonical_root();
+
+        if canonical_root != head_block.message().parent_root() {
```
Comment (pawanjay176) [inline beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs:120]: Could we also get an update message where `update.attested_header` references a block which we have already seen? In this case, would it be better to simply check if we have seen `canonical_root` in fork choice instead of checking if it's equal to current head's parent? Might be completely wrong on this one, so please feel free to correct me :)

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs [hunk 2]
```diff
@@ -473,9 +527,52 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
...
+                if self.lc_updates_delay_queue.len() >= MAXIMUM_QUEUED_LIGHT_CLIENT_UPDATES {
+                    if self.lc_update_delay_debounce.elapsed() {
+                        error!(
+                            log,
+                            "Light client updates delay queue is full";
+                            "queue_size" => MAXIMUM_QUEUED_LIGHT_CLIENT_UPDATES,
+                            "msg" => "check system clock"
+                        );
+                    }
+                    // Drop the attestation.
```
Comment (pawanjay176) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]:
```suggestion
                    // Drop the light client update.
```

File: beacon_node/network/src/beacon_processor/mod.rs [hunk 3]
```diff
@@ -694,6 +700,20 @@ impl<T: BeaconChainTypes> std::convert::From<ReadyWork<T>> for WorkEvent<T> {
                     seen_timestamp,
                 },
             },
+            ReadyWork::LCUpdate(QueuedLCUpdate {
```
Comment (pawanjay176) [inline beacon_node/network/src/beacon_processor/mod.rs]: I think it would be nice to use the full name here specially for the public structs and enums. 
`ReadyWork::LightClientUpdate`, `QueuedLightClientUpdate` etc

**Review by pawanjay176 — COMMENTED** [review]
Looking good overall. Just a few minor nits

File: beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs (hunk 1 above)
Comment (GeemoCandama) [reply beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs:120]: I'm not certain. I don't think that would happen because the only time new optimistic updates are supposed to be published to pubsub is: "Whenever fork choice selects a new head block with a sync aggregate participation >= MIN_SYNC_COMMITTEE_PARTICIPANTS and a post-Altair parent block, full nodes with at least one validator assigned to the current sync committee at the block's slot". "If attested_header.slot increased, a LightClientOptimisticUpdate SHOULD be broadcasted to the pubsub topic light_client_optimistic_update if no matching message has not yet been forwarded as part of gossip validation." 
 https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#sync-committee . That's where my understanding comes from but I might be thinking about it incorrectly.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs [hunk 4]
```diff
@@ -511,12 +612,58 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
...
                     }
                 }
+                // Unqueue the light client optimistic updates we have for this root, if any.
+                if let Some(queued_lc_id) = self
+                    .awaiting_lc_updates_per_parent_root
+                    .remove(&parent_root)
+                {
+                    debug!(
+                        log,
+                        "unqueue light client optimistic updates";
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]:
```suggestion
                        "Dequeuing light client optimistic updates";
```

File: beacon_node/network/src/beacon_processor/worker/gossip_methods.rs [hunk 5]
```diff
@@ -1377,22 +1381,76 @@ impl<T: BeaconChainTypes> Worker<T> {
...
+                            if sender.try_send(msg).is_err() {
+                                error!(
+                                    self.log,
+                                    "Failed to send optimistic update for re-processing";
+                                )
+                            }
+                        } else {
+                            debug!(
+                                self.log,
+                                "Not sendinglight clientupdate because it had been reprocessed";
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/worker/gossip_methods.rs]:
```suggestion
                                "Not sending light client update because it had been reprocessed";
```

File: beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs:120]: I think the only time this might be faulty is if there are re-orgs near the head. Our current approach to verification always uses the canonical head, so it will reject updates from competing non-canonical chains that have been forked for longer than 1 block.

We initially started using the canonical head block because we thought it would help with caching, but it's actually useless, so perhaps we can refactor (at some point) to load _a_ block from `update.signature_slot`.

TL;DR: I think this is OK as-is, but might need revisiting in future.

File: beacon_node/network/src/beacon_processor/worker/gossip_methods.rs [hunk 6]
```diff
@@ -1377,22 +1381,76 @@ impl<T: BeaconChainTypes> Worker<T> {
...
+                            "Optimistic update for unknown block";
+                            "peer_id" => %peer_id,
+                            "parent_root" => ?parent_root
+                        );
+
+                        if let Some(sender) = reprocess_tx {
+                            let msg = ReprocessQueueMessage::UnknownLightClientOptimisticUpdate(
+                                QueuedLightClientUpdate {
+                                    peer_id,
+                                    message_id,
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/worker/gossip_methods.rs:1418]: I think it would be a nice small improvement to plumb through the `parent_root` here so we don't need to recompute it later.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs [hunk 7]
```diff
@@ -473,9 +529,54 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
...
+                    .lc_updates_delay_queue
+                    .insert(lc_id, QUEUED_LIGHT_CLIENT_UPDATE_DELAY);
+
+                // Register the light client update for the corresponding root.
+                self.awaiting_lc_updates_per_parent_root
+                    .entry(
+                        queued_light_client_optimistic_update
+                            .light_client_optimistic_update
+                            .attested_header
+                            .canonical_root(),
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: Rather than recomputing the root here, we could use the `parent_root` (added to `QueuedLightClientUpdate`).

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs [hunk 8]
```diff
@@ -511,12 +612,58 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
...
+                                    delay_key,
+                                )
+                            },
+                        ) {
+                            // Remove the delay
+                            self.lc_updates_delay_queue.remove(&delay_key);
+
+                            // Send the work
+                            match self.ready_work_tx.try_send(work) {
+                                Ok(_) => debug!(log, "reprocessing light client update sent"),
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: I think we probably shouldn't bother logging each message here individually. Even debug level logs get saved to disk by default. We could change the level to trace and/or add a `"count" => queued_lc_id.len()` to the `debug!` log above.

File: beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs [hunk 9]
```diff
@@ -591,6 +738,41 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
...
+                    &metrics::BEACON_PROCESSOR_REPROCESSING_QUEUE_EXPIRED_OPTIMISTIC_UPDATES,
+                );
+
+                if let Some((parent_root, work)) = self.queued_lc_updates.remove(&queued_id).map(
+                    |(queued_lc_update, _delay_key)| {
+                        (
+                            queued_lc_update
+                                .light_client_optimistic_update
+                                .attested_header
+                                .canonical_root(),
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs]: This could also use the cached `parent_root` if we had it stored

**Review by michaelsproul — COMMENTED** [review]
Fantastic!

Just a few tweaks and then I'm happy to merge it!

**Review by michaelsproul — APPROVED** [review]
YES!
